kernel/kernel-defaults/src/main/java/io/delta/kernel/defaults/internal/DefaultKernelUtils.java

```diff
@@ -87,16 +123,55 @@ private static List<Type> pruneFields(GroupType type, StructType deltaDataType)
             .collect(Collectors.toList());
     }
 
-    private static Type prunedType(Type type, DataType deltaType) {
+    private static Type prunedType(Type type, DataType deltaType, boolean hasFieldIds) {
         if (type instanceof GroupType && deltaType instanceof StructType) {
             GroupType groupType = (GroupType) type;
             StructType structType = (StructType) deltaType;
-            return groupType.withNewFields(pruneFields(groupType, structType));
+            return groupType.withNewFields(pruneFields(groupType, structType, hasFieldIds));
         } else {
             return type;
         }
     }
 
+    /**
+     * Recursively checks whether the given data type has any Parquet field ids in it.
```

Comment:
Shouldn't the presence/absence of (usable) field ids be controlled by the column mapping table feature? At best, this method should be for validating that the schema provides field ids always and only when it should, but that doesn't seem to be how the code is currently structured.

Reply:
If we did pass an arg, it would just a boolean `usesFieldIds` I think? (column mapping is a Delta issue the parquet handler shouldn't have to understand). 

But that's an interesting point. I don't know the right answer. It just seemed annoying/expensive if we have to search the entire schema (which could be both wide and deep)?

Reply:
I agree it is not efficient to go through the schema to find if it has any ids. Btw Spark Parquet reader uses a config and also searches the parquet schema for ids to figure out whether it needs to do id based column mapping